Sphinx dev docs fixes

peterfpeterson opened the pull request:
**Description of work.**
 [`master_clean-rhel7`](https://builds.mantidproject.org/view/Master%20Pipeline/job/master_clean-rhel7/1672/) is failing due to not launching sphinx correctly for building the developer documentation. Since this has already been sorted out for the user documentation, centralize the behavior and use it in both places. There are also some an update for numpy intersphinx and some minor mark-up changes.

**To test:**

```sh
ninja dev-docs-html docs-html
```
should complete without warnings.

*There is no associated issue.*

*This does not require release notes* because it is repairing configuration of the developer documentation builds.

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

peterfpeterson committed:
Fix numpy intersphinx

peterfpeterson committed:
Fix sphinx warnings

peterfpeterson committed:
Consolidate running sphinx to a single method

martyngigg commented on dev-docs/CMakeLists.txt:
```diff
@@ -7,23 +7,14 @@ set(DOCTREE_DIR ${CMAKE_CURRENT_BINARY_DIR}/doctree)
+add_custom_target(dev-docs-${BUILDER}
+                  COMMAND ${Python_EXECUTABLE} -m ${SPHINX_MAIN} ${SPHINX_NO_COLOR}
+                          -w dev_docs_warnings.txt
+                          -W --keep-going -b ${BUILDER}
```
I'm getting an error in my CentOS 7 docker container:
```
[1/1] Building html developer documentation
FAILED: dev-docs/CMakeFiles/dev-docs-html
cd /mantid_build/dev-docs && /usr/bin/python3.6 -m sphinx.__init__ -w dev_docs_warnings.txt -W --keep-going -b html -d /mantid_build/dev-docs/doctree /mantid_src/dev-docs/source /mantid_build/dev-docs/html
Error: option --keep-going not recognized
```
It looks like this `--keep-going` argument only got added in [Sphinx v2.0](https://github.com/sphinx-doc/sphinx/commit/925bc187eacbc0fbdd2c56f360a040a23cb13145) so I think we need to similar protection to `--no-color`

peterfpeterson committed:
Make more options common between builds

--keep-going isn't available until sphinx 2.0
-W turns warnings into errors which should help jenkins figure out that
the build finished with warnings

martyngigg reviewed (commented):
More problems furhter down I'm afraid:
```
[abc@ddce46d9afdf mantid_build]$ ninja dev-docs-html
[1/1] Building html developer documentation
FAILED: dev-docs/CMakeFiles/dev-docs-html
cd /mantid_build/dev-docs && /usr/bin/python3.6 -m sphinx.__init__ -w dev_docs_warnings.txt -W -b html -d /mantid_build/dev-docs/doctree /mantid_src/dev-docs/source /mantid_build/dev-docs/html
Making output directory...
Running Sphinx v1.2.3

Exception occurred:
  File "conf.py", line 17, in setup
AttributeError: 'Sphinx' object has no attribute 'add_css_file'
The full traceback has been saved in /tmp/sphinx-err-ivt7yjte.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://bitbucket.org/birkenfeld/sphinx/issues/>. Thanks!
ninja: build stopped: subcommand failed.
```

Maybe we need this `if` clause in `dev-docs/source/conf.py: https://github.com/mantidproject/mantid/blob/f157ce380e414ed8ef88664ae5a1ae49f60b1f57/docs/source/conf.py#L41

I'll be glad when conda is sorted so we can stop having this kind of thing.

peterfpeterson committed:
Be more flexible on how the stylesheet is added

peterfpeterson committed:
Turn off -W option due to intersphinx

martyngigg reviewed (approved):
This latest version works. We sacrifice treating warning as errors but we don't do that with compiler warnings so it seems overkill to do it with docs warnings.
